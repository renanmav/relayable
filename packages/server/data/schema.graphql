input AcceptAnswerInput {
  id: ID!
  clientMutationId: String
}

type AcceptAnswerPayload {
  error: String
  answer: Answer
  clientMutationId: String
}

"""Answer data"""
type Answer implements Node & Post {
  """The ID of an object"""
  id: ID!
  _id: String
  question: Question
  content: String
  upvotes: Int
  downvotes: Int
  is_accepted: Boolean
  author: User
  createdAt: String
  updatedAt: String
}

input CreateAnswerInput {
  """ID of the question"""
  question: ID!
  content: String!
  clientMutationId: String
}

type CreateAnswerPayload {
  error: String
  answer: Answer
  clientMutationId: String
}

input CreateQuestionInput {
  title: String!
  content: String!
  tags: [String]
  clientMutationId: String
}

type CreateQuestionPayload {
  error: String
  question: Question
  clientMutationId: String
}

input DeleteQuestionInput {
  id: ID!
  clientMutationId: String
}

type DeleteQuestionPayload {
  error: String
  clientMutationId: String
}

input EditAnswerInput {
  id: ID!
  content: String!
  clientMutationId: String
}

type EditAnswerPayload {
  error: String
  answer: Answer
  clientMutationId: String
}

input EditQuestionInput {
  id: ID!
  title: String
  content: String
  tags: [String]
  clientMutationId: String
}

type EditQuestionPayload {
  error: String
  question: Question
  clientMutationId: String
}

input LoginWithGithubInput {
  code: String!
  clientMutationId: String
}

type LoginWithGithubPayload {
  token: String
  user: User
  clientMutationId: String
}

type Mutation {
  LoginWithGithub(input: LoginWithGithubInput!): LoginWithGithubPayload

  """Use this mutation to create a question"""
  CreateQuestion(input: CreateQuestionInput!): CreateQuestionPayload

  """Mutation to edit the question if logged in user is the author"""
  EditQuestion(input: EditQuestionInput!): EditQuestionPayload

  """Use this mutation to delete a question if you owns it"""
  DeleteQuestion(input: DeleteQuestionInput!): DeleteQuestionPayload

  """
  Mutation to view the question
  Each user represents one view
  """
  ViewQuestion(input: ViewQuestionInput!): ViewQuestionPayload

  """
  Mutation to up/down vote a question
  
  Each user vote only once (and can change his vote)
  """
  VoteQuestion(input: VoteQuestionInput!): VoteQuestionPayload

  """Use this mutation to answer a question"""
  CreateAnswer(input: CreateAnswerInput!): CreateAnswerPayload

  """Mutation to edit the answer if logged in user is the author"""
  EditAnswer(input: EditAnswerInput!): EditAnswerPayload

  """
  Mutation to up/down vote a answer
  
  Each user vote only once (and can change his vote)
  """
  VoteAnswer(input: VoteAnswerInput!): VoteAnswerPayload

  """
  Mutation to accept the answer
  Only the owner of the question can accept the answer
  Only one answer can be accepted
  """
  AcceptAnswer(input: AcceptAnswerInput!): AcceptAnswerPayload
}

type NewQuestionPayload {
  edge(token: String!): QuestionEdge
}

type NewViewPayload {
  edge(token: String!): QuestionEdge
}

type NewVotePayload {
  edge(token: String!): QuestionEdge
}

"""An object with an ID"""
interface Node {
  """The id of the object."""
  id: ID!
}

"""Information about pagination in a connection."""
type PageInfo {
  """When paginating forwards, are there more items?"""
  hasNextPage: Boolean!

  """When paginating backwards, are there more items?"""
  hasPreviousPage: Boolean!

  """When paginating backwards, the cursor to continue."""
  startCursor: String

  """When paginating forwards, the cursor to continue."""
  endCursor: String
}

"""Post interface to enforce some fields"""
interface Post {
  content: String
  upvotes: Int
  downvotes: Int
  author: User
  createdAt: String
  updatedAt: String
}

"""The root of all queries"""
type Query {
  """Use this query to fetch GitHub login URL"""
  githubLoginUrl: String

  """Fetches an object given its ID"""
  node(
    """The ID of an object"""
    id: ID!
  ): Node

  """Get your own data based on provided Authorization token"""
  me: User

  """Fetch a user by its login"""
  user(login: String!): User

  """
  Returns a connection with connectionArgs and/or the Author global ID
  
  You can search questions too with the search arg
  
  Search its done by title, content and tags fields
  """
  questions(after: String, first: Int, before: String, last: Int, authorId: ID, search: String): QuestionConnection

  """
  Returns the average difference between a question creation and its first answer
  
  Value returned is in minutes
  """
  questionAvgResponse: String
}

"""Question data"""
type Question implements Node & Post {
  """The ID of an object"""
  id: ID!
  _id: String
  title: String
  content: String
  upvotes: Int
  downvotes: Int
  views: Int
  anonymous_views: Int
  total_views: Int
  tags: [String]
  author: User
  answers: [Answer]
  createdAt: String
  updatedAt: String
}

"""A connection to a list of items."""
type QuestionConnection {
  """Information to aid in pagination."""
  pageInfo: PageInfo!

  """A list of edges."""
  edges: [QuestionEdge]
}

"""An edge in a connection."""
type QuestionEdge {
  """The item at the end of the edge"""
  node: Question

  """A cursor for use in pagination"""
  cursor: String!
}

type Subscription {
  NewQuestion: NewQuestionPayload
  NewView: NewViewPayload
  NewVote: NewVotePayload
}

"""User data"""
type User implements Node {
  """The ID of an object"""
  id: ID!
  _id: String
  github_id: String
  name: String
  login: String
  avatar_url: String
}

input ViewQuestionInput {
  id: ID!
  clientMutationId: String
}

type ViewQuestionPayload {
  error: String
  question: Question
  clientMutationId: String
}

input VoteAnswerInput {
  id: ID!
  up: Boolean
  down: Boolean
  clientMutationId: String
}

type VoteAnswerPayload {
  error: String
  answer: Answer
  clientMutationId: String
}

input VoteQuestionInput {
  id: ID!
  up: Boolean
  down: Boolean
  clientMutationId: String
}

type VoteQuestionPayload {
  error: String
  question: Question
  clientMutationId: String
}
