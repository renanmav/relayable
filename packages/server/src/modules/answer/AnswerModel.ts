import mongoose, { Schema, Document, Model } from 'mongoose'

import { IUser } from '../user/UserModel'
import { IQuestion } from '../question/QuestionModel'

const answerSchema = new Schema(
  {
    question: {
      type: Schema.Types.ObjectId,
      ref: 'Question',
    },
    content: {
      type: String,
      required: true,
    },
    upvotes: [
      {
        type: Schema.Types.ObjectId,
        ref: 'User',
      },
    ],
    downvotes: [
      {
        type: Schema.Types.ObjectId,
        ref: 'User',
      },
    ],
    author: {
      type: Schema.Types.ObjectId,
      ref: 'User',
      required: true,
    },
    is_accepted: {
      type: Boolean,
      default: false,
    },
  },
  {
    timestamps: {
      createdAt: 'createdAt',
      updatedAt: 'updatedAt',
    },
    collection: 'answer',
  }
)

export interface IAnswer extends Document {
  question: IQuestion | string
  content: string
  upvotes: IUser[]
  downvotes: IUser[]
  author: IUser | string
  is_accepted: boolean
  createdAt: {
    toISOString: () => string
    toString: () => string
  }
  updatedAt: {
    toISOString: () => string
    toString: () => string
  }
}

const AnswerModel: Model<IAnswer> = mongoose.model('Answer', answerSchema)

export default AnswerModel
