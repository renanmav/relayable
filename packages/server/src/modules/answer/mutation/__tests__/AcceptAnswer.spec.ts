import { graphql } from 'graphql'
import { toGlobalId } from 'graphql-relay'

import {
  getContext,
  createRows,
  connectMongoose,
  clearDbAndRestartCounters,
  disconnectMongoose,
} from '../../../../../test/helper'
import { schema } from '../../../../schema'
import AnswerModel from '../../AnswerModel'

beforeAll(connectMongoose)

beforeEach(clearDbAndRestartCounters)

afterAll(disconnectMongoose)

const gql = String.raw

const query = gql`
  mutation acceptAnswer($id: ID!) {
    AcceptAnswer(input: { id: $id }) {
      error
      answer {
        is_accepted
      }
    }
  }
`
const rootValue = {}

it("should not accept the answer if isn't question owner", async () => {
  const user = await createRows.createUser()
  const question = await createRows.createQuestion({ author: user._id })

  const otherUser = await createRows.createUser()
  const answer = await createRows.createAnswer(question, {
    author: otherUser._id,
  })

  const context = getContext({ user: otherUser })
  const variables = {
    id: toGlobalId('Answer', answer._id),
  }

  const result = await graphql(schema, query, rootValue, context, variables)

  expect(result).toMatchSnapshot()
})

it('should accept only one answer', async () => {
  const user = await createRows.createUser()
  const question = await createRows.createQuestion({ author: user._id })

  const answerUser = await createRows.createUser()
  const answer = await createRows.createAnswer(question, {
    author: answerUser._id,
    is_accepted: true,
  })
  const acceptedAnswer = await createRows.createAnswer(question, {
    author: answerUser._id,
  })

  const context = getContext({ user })
  const variables = {
    id: toGlobalId('Answer', acceptedAnswer._id),
  }

  const result = await graphql(schema, query, rootValue, context, variables)

  const _answer = await AnswerModel.findById(answer._id)

  expect(result).toMatchSnapshot()
  expect(_answer!.is_accepted).toBe(false)
})
