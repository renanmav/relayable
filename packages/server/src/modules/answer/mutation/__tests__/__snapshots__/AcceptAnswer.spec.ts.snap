// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should accept only one answer 1`] = `
Object {
  "data": Object {
    "AcceptAnswer": Object {
      "answer": Object {
        "is_accepted": true,
      },
      "error": null,
    },
  },
}
`;

exports[`should not accept the answer if isn't question owner 1`] = `
Object {
  "data": Object {
    "AcceptAnswer": Object {
      "answer": null,
      "error": "You don't own this question to accept this answer",
    },
  },
}
`;
